### PR TITLE
Neighborfiles util

### PR DESCRIPTION
**Pull request by Will-ShaoHua**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)

## Checklist
- [ ] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-jetbrains/blob/master/CONTRIBUTING.md#contributing-via-pull-requests)** if the change is customer-facing in the IDE.
- [ ] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt [hunk 1]
```diff
@@ -165,6 +167,71 @@ abstract class CodeWhispererFileCrawler : FileCrawler {
...
+    private fun search(psiFile: PsiFile, distance: Int): Set<PsiFile> = runReadAction {
+        search(psiFile.containingDirectory, distance, true) +
+            search(psiFile.containingDirectory, distance, false)
+    }
+
+    private fun search(psiDir: PsiDirectory, distance: Int, goDown: Boolean): Set<PsiFile> {
+        var d = distance
+        val res = mutableListOf<PsiFile>()
+        val pendingVisit = mutableListOf(psiDir)
+        val visisted = mutableMapOf(psiDir to false)
```
Comment (andrewyuq) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt:204]: Typo

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt [hunk 2]
```diff
@@ -165,6 +167,71 @@ abstract class CodeWhispererFileCrawler : FileCrawler {
...
+
+                val dirs = if (goDown) {
+                    dir.subdirectories.toList()
+                } else {
+                    dir.parentDirectory?.let {
+                        listOf(it)
+                    }.orEmpty()
+                }
+
+                toVisit = dirs
```
Comment (andrewyuq) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt:224]: Feel like tovisit should addAll here not equal

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt (hunk 1 above)
Comment (Will-ShaoHua) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt]: updated

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt (hunk 2 above)
Comment (Will-ShaoHua) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt]: i think either works, but yes addAll should be the better one.

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt [hunk 3]
```diff
@@ -165,6 +167,71 @@ abstract class CodeWhispererFileCrawler : FileCrawler {
...
+     */
+    fun neighborFiles(psiFile: PsiFile): Set<PsiFile> = search(psiFile, NEIGHBOR_FILES_DISTANCE).filterNot { it == psiFile }.toSet()
+
+    private fun search(psiFile: PsiFile, distance: Int): Set<PsiFile> = runReadAction {
+        search(psiFile.containingDirectory, distance, true) +
+            search(psiFile.containingDirectory, distance, false)
+    }
+
+    private fun search(psiDir: PsiDirectory, distance: Int, goDown: Boolean): Set<PsiFile> {
+        var d = distance
```
Comment (andrewyuq) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt:201]: distance is Int which is pass by value so it doesn't really need a `var d` here.

**Review by andrewyuq — APPROVED** [review]

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt [hunk 4]
```diff
@@ -165,6 +167,71 @@ abstract class CodeWhispererFileCrawler : FileCrawler {
...
+                    dir.parentDirectory?.let {
+                        listOf(it)
+                    }.orEmpty()
+                }
+
+                toVisit.addAll(dirs)
+                visited[dir] = true
+            }
+
+            pendingVisit.addAll(toVisit)
```
Comment (andrewyuq) [inline plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt:228]: pendingVisit can just = toVisit, so you don't need to foreach the already visited ones(front-loaded ones) again and again.

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt (hunk 3 above)
Comment (Will-ShaoHua) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt:201]: yea but i prefer not touching func argument directly so i make it a different var

File: plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt (hunk 4 above)
Comment (Will-ShaoHua) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt:228]: yea makes sense to me

Comment (Will-ShaoHua) [reply plugins/amazonq/codewhisperer/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codewhisperer/util/CodeWhispererFileCrawler.kt:228]: that's a good callout, thanks andrew

**Pull request merged**